From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: corymhall <43035978+corymhall@users.noreply.github.com>
Date: Fri, 18 Apr 2025 12:51:25 -0400
Subject: [PATCH] Enhancements to shim

This patch combines a couple of patches that all effect shim.go

- move-shim-logic-to-upstream-as-a-patch.patch
- Optimize-startup-performance.patch
- Fix-markTagsAllNotComputedForResources-to-recognize-.patch
- DisableTagSchemaCheck-for-PF-provider.patch

diff --git a/shim/shim.go b/shim/shim.go
new file mode 100644
index 0000000000..858599b33f
--- /dev/null
+++ b/shim/shim.go
@@ -0,0 +1,40 @@
+package shim
+
+import (
+	"context"
+
+	pfprovider "github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-aws/internal/provider"
+	"github.com/hashicorp/terraform-provider-aws/internal/provider/fwprovider"
+	"github.com/hashicorp/terraform-provider-aws/internal/tags"
+)
+
+type UpstreamProvider struct {
+	SDKV2Provider           *schema.Provider
+	PluginFrameworkProvider pfprovider.Provider
+}
+
+func NewUpstreamProvider(ctx context.Context) (UpstreamProvider, error) {
+	primary, err := provider.New(ctx)
+	if err != nil {
+		return UpstreamProvider{}, err
+	}
+	pf, err := fwprovider.New(ctx, primary)
+	if err != nil {
+		//lintignore:R009
+		panic(err)
+	}
+	return UpstreamProvider{
+		SDKV2Provider:           primary,
+		PluginFrameworkProvider: pf,
+	}, nil
+}
+
+type TagConfig = tags.DefaultConfig
+
+type TagIgnoreConfig = tags.IgnoreConfig
+
+func NewTagConfig(ctx context.Context, i interface{}) TagConfig {
+	return TagConfig{Tags: tags.New(ctx, i)}
+}
